maven-caching/src/main/java/org/apache/maven/caching/CacheControllerImpl.java

```diff
@@ -397,67 +415,126 @@ private RestoredArtifact restoredArtifact( org.apache.maven.artifact.Artifact pa
+                        continue;
+                    }
+
+                    if ( JarFile.MANIFEST_NAME.equals( entryName ) )
+                    {
+                        replaceEntry( jarFile, entry, originalImplVersion, implVersion, jos );
+                        continue;
+                    }
+                    jos.putNextEntry( entry );
+                    try ( InputStream entryInputStream = jarFile.getInputStream( entry ) )
+                    {
+                        int bytesRead;
+                        while ( ( bytesRead = entryInputStream.read( buffer ) ) != -1 )
+                        {
+                            jos.write( buffer, 0, bytesRead );
+                        }
+                    }
+                }
             }
-            LOGGER.debug( "Downloaded artifact " + artifact.getArtifactId() + " to: " + artifactFile );
-            return artifactFile.toFile();
-        } );
-        if ( !cacheConfig.isLazyRestore() )
-        {
-            downloadTask.run();
-            downloadTask.get();
         }
-        return downloadTask;
+        return tmpJarFile.toPath();
     }
 
-    private void putChecksum( Artifact artifact, String projectChecksum )
+    private static void replaceEntry( JarFile jarFile, JarEntry entry,
+                                      String toReplace, String replacement, JarOutputStream jos ) throws IOException
     {
-        final DigestItem projectArtifact = DtoUtils.createdDigestedByProjectChecksum( artifact, projectChecksum );
-        final String dependencyKey = KeyUtils.getArtifactKey( artifact );
-        artifactDigestByKey.put( dependencyKey, projectArtifact );
+        String fullManifest = IOUtils.toString( jarFile.getInputStream( entry ), StandardCharsets.UTF_8.name() );
+        String modified = fullManifest.replaceAll( toReplace, replacement );
 
-        final ArtifactHandler artifactHandler = artifactHandlerManager.getArtifactHandler( artifact.getType() );
-        String packaging = artifactHandler.getPackaging();
-        if ( !Objects.equals( packaging, artifact.getType() ) )
-        {
-            artifactDigestByKey.put( KeyUtils.getArtifactKey( artifact, packaging ), projectArtifact );
-        }
-        String extension = artifactHandler.getExtension();
-        if ( !Objects.equals( extension, artifact.getType() ) && !Objects.equals( extension, packaging ) )
+        byte[] bytes = modified.getBytes( StandardCharsets.UTF_8 );
+        JarEntry newEntry = new JarEntry( entry.getName() );
+        jos.putNextEntry( newEntry );
+        jos.write( bytes );
+    }
+
+    private static boolean isArchive( File file )
+    {
+        String fileName = file.getName();
+        if ( !file.isFile() || file.isHidden() )
         {
-            artifactDigestByKey.put( KeyUtils.getArtifactKey( artifact, extension ), projectArtifact );
+            return false;
         }
+        return fileName.endsWith( ".jar" ) || fileName.endsWith( ".zip" )
```

Comment:
is [StringUtils#endsWithAny](https://commons.apache.org/proper/commons-lang/apidocs/org/apache/commons/lang3/StringUtils.html#endsWithAny-java.lang.CharSequence-java.lang.CharSequence...-) avilable? 
```
 StringUtils.endsWithAny("abcxyz", new String[] {null, "xyz", "abc"}) = true
 StringUtils.endsWithAny("abcXYZ", "def", "XYZ") = true
 StringUtils.endsWithAny("abcXYZ", "def", "xyz") = false
```

Reply:
> StringUtils#endsWithAny could be handy here

changed